[BUG] minirocket: zero division errors #5174

benshaw2 opened the pull request:
#### Reference Issues/PRs

Fixes #5174.


#### What does this implement/fix? Explain your changes.

MultiRocket and MiniRocket transforms are not well defined when the parameter num_kernels is less than 84.

The reason at least 84 kernels must be chosen is as follows. Multi/mini rocket transforms rely on precisely 84 "base" kernels. Kernels used to transform the data are generated as dilations of the base 84. When a user specifies the parameter num_kernels, the algorithm generates a number of kernels equal to the nearest multiple of 84 (rounding down), since the number of dilations per kernel is num_kernels // 84. If the parameter num_kernels is less than 84, no transformation kernels can be generated, since the number of dilations per kernel is 0.

The authors did not foresee this issue, since they always set the parameter num_kernels to be in the thousands and recommended this practice to subsequent users.

When the num_kernels is less than 84 for multi/mini rocket transforms, the actual number of kernels used is set to 84. A new attribute called num_kernels_ gives the actual number of kernels to be used. For transforms, it is set during _fit. For classifiers and regressors using the "minirocket" or "multirocket" transforms, it is calculated at initialization.


#### Does your contribution introduce a new dependency? If yes, which one?

No.


#### What should a reviewer concentrate their feedback on?

Does the divide by zero error happen for multirocket or minirocket transforms? Is the num_kernels_ attribute accessible?

#### Did you add any tests for the change?

In _rocket_classifier.py and _rocket_regressor.py, some failing parameter values were added to get_test_params: when the parameter rocket_transform is either "minirocket" or "multirocket" and num_kernels < 84, there should no longer be a division by zero error.


#### Any other comments?

No other comments.


#### PR checklist

Please go through the checklist below. Please feel free to remove points if they are not applicable.


##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

<!--
Thanks for contributing!
-->

fkiraly reviewed (changes requested):
Thanks!

We need to make sure that there is a test that checks that the failure is no longer there.

For this, could you add the failing parameter combinations to `get_test_params`, in each estimator?

fkiraly commented:
Also, could you kindly explain why the minimum is 84 specifically?
If that is a hard bound, it should be mentioned in the docstrings.

benshaw2 commented:
Thanks @fkiraly for the feedback. I have added some failing parameters to get_test_params in the classifier and regressor python files: is this what is needed? I mentioned in each docstring that num_kernels must be at least 84 for minirocket and multirocket transforms, and I edited the original PR request comment to explain why this must be. I also added my github information to the .all-contributorsrc file. Let me know if anything else needs changing or if anything else should be added.

fkiraly commented on sktime/classification/kernel_based/_rocket_classifier.py:
```diff
@@ -136,6 +138,15 @@ def __init__(
         self.num_kernels = num_kernels
         self.rocket_transform = rocket_transform
+        
+        if num_kernels < 84:
```
simpler: `if num_kernels < 84 and rocket_transform in ["multirocket", "minirocket"]`

fkiraly reviewed (changes requested):
> The reason at least 84 kernels must be chosen is as follows. Multi/mini rocket transforms rely on precisely 84 "base" kernels. Kernels used to transform the data are generated as dilations of the base 84. When a user specifies the parameter num_kernels, the algorithm generates a number of kernels equal to the nearest multiple of 84 (rounding down), since the number of dilations per kernel is num_kernels // 84. If the parameter num_kernels is less than 84, no transformation kernels can be generated, since the number of dilations per kernel is 0.

Oh, I see. So the "effective number of kernels" is always a multiple of 84.

Then, may I suggest an alternative resolution:

* say in the docstring that number of kernels should be a multiple of 84, otherwise it will be rounded down to the next positive multiple of 84, or up to 84 if lower.
* This way, too small parameters will lead to 84.
* I would also add a `num_kernels_` fitted parameter, written to `self` in `_fit`, which is the actual number of kernels used, so users can check. This should be explained in the `Attributes` section.

This way, passing too small parameters should not cause an error, making the estimator more robust. Of course it needs to be documented that it *always* gets rounded.

fkiraly commented on sktime/classification/kernel_based/_rocket_classifier.py:
```diff
@@ -236,4 +247,4 @@ def get_test_params(cls, parameter_set="default"):
             return {"num_kernels": 100}
         else:
-            return {"num_kernels": 20}
+            return {"num_kernels": 20, "rocket_transform": "minirocket"}
```
this should actually be a lits of multiple dictionaries, each being a parameter set that is tested.

ou can also omit the `else` for better readability, as it follows an `if`-`return` statement. If anything past the `if` is reached, we know it is `else` (logically)

benshaw2 commented:
Thanks @fkiraly for the feedback. I have changed the resolution strategy so that no errors are returned at all: when num_kernels is less than 84 for the multirocket or minirocket transform, num_kernels is effectively set to 84. I have also introduced the new num_kernels_ attribute and updated the documentation. For the actual transforms, the num_kernels_ attribute is set during _fit, though for the classifier and regressor using rocket, the attribute is set at initialization. I also believe I addressed your other comments, though once again let me know what other changes need to be made.

fkiraly commented on sktime/classification/kernel_based/_rocket_classifier.py:
```diff
@@ -233,7 +249,8 @@ def get_test_params(cls, parameter_set="default"):
             instance.
             ``create_test_instance`` uses the first (or only) dictionary in ``params``.
         """
-        if parameter_set == "results_comparison":
```
we should leave the `results_comparison` parameter set here, this is used in consistency experiments.

I.e., there would be an "if", but no "else".

fkiraly commented on sktime/transformations/panel/rocket/_minirocket.py:
```diff
@@ -85,14 +95,14 @@ class MiniRocket(BaseTransformer):
     def __init__(
         self,
-        num_kernels=10_000,
+        num_kernels=9_996,
```
we need to leave it at 10.000 as otherwise it may break user code that uses `get_params`.

fkiraly reviewed (changes requested):
Great!

I think we can merge this, up to small changes due to the fact that we cannot change standing interfaces (parameter defaults and "comparison" parameter sets), without lengthy deprecation. I would just leave them as they are:

* no change to init defaults
* no removal of the `results_comparison` parameter set

We could also go through the deprecation process https://www.sktime.net/en/stable/developer_guide/deprecation.html if we really want to make the changes, though I feel that's maybe not really worth the effort here.

The overarching reason is "no user code should break without warning if we can avoid it".

benshaw2 commented:
I've now changed the init defaults back and restored the results_comparison parameter set in _rocket_classifier.py. If that's it, great! Otherwise, let me know of any other changes that need to be made. Thanks.

fkiraly reviewed (approved):
That's it & great!
